server/routes/utils/adHelpers.ts

```diff
@@ -436,3 +526,59 @@ export const getLatestDetectorTasksQuery = () => {
 export const isRealTimeTask = (task: any) => {
   return get(task, 'task_type', '').includes(REALTIME_TASK_TYPE_PREFIX);
 };
+
+export const getLatestTasksForDetectorQuery = (detectorId: string) => {
+  return {
+    size: 3,
+    query: {
+      bool: {
+        filter: [
+          {
+            term: {
+              detector_id: detectorId,
+            },
+          },
+          {
+            term: {
+              is_latest: 'true',
+            },
+          },
+          {
+            terms: {
+              task_type: [
+                'REALTIME_HC_DETECTOR',
+                'REALTIME_SINGLE_ENTITY',
+                'HISTORICAL_SINGLE_ENTITY',
+                'HISTORICAL_HC_DETECTOR',
+                'HISTORICAL',
+              ],
+            },
+          },
+        ],
+      },
+    },
+  };
+};
+
+// Helper fn to get the latest realtime task for some detector.
+// It is possible that the backend returns multiple realtime tasks
+// set to 'is_latest=true' when backfilling legacy realtime detectors.
+export const getLatestRealtimeTask = (taskList: any[]) => {
+  const realtimeTaskList = taskList.filter((task: any) => isRealTimeTask(task));
+  let realtimeTask = get(realtimeTaskList, 0);
+  if (get(realtimeTaskList, 'length', 0) > 1) {
+    realtimeTask =
+      get(realtimeTaskList, '0.execution_start_time') >
```

Comment:
Can we just sort by `execution_start_time` and get the first realtime task?

Reply:
Since the detectors can return 3 possible tasks (1 historical, 2 realtime), it is possible that a historical task is lost if the 2 realtime tasks both have newer `execution_start_time`s. We could sort if we did separate queries for realtime and historical, but prefer to keep as one. Let me know what you think.

Reply:
Recently learned that during the backfilling process on backend, that multiple realtime tasks can be created during some edge cases as new nodes join the cluster (multiple nodes join at the same time and see that no realtime tasks exist for some detector, so both create them). And even more recently learned that it may not be just 2, but an arbitrary amount, if x number of nodes join the cluster and trigger this edge case.

Because of this, I'm thinking of changing to Yaliang's suggestion of having a separate query to fetch realtime tasks, sort by `execution_start_time`, and return the top 1. But not sure if the performance of making 2 search task API calls vs. 1 call is worth the performance hit, since it is just for handling this edge case.

Reply:
I'm not sure about the how likely the edge case will happen, but it will happen if some new node starts first, then it start to backfill ,then other nodes join cluster, the AD version hash ring may change, so other node will backfill again. It will be safe to query separately for realtime and historical.

Reply:
Got it. Will change to have separate queries.

Reply:
@kaituo @ylwu-amzn changed to use separate queries in latest commit.